### PR TITLE
stub: detect invalid states on server side (eg zero responses for unary)

### LINKED ISSUE #2243: Detect zero-message response to unary request on server-side
It [appears](https://github.com/grpc/grpc-java/blob/v1.0.0/core/src/main/java/io/grpc/internal/ServerCallImpl.java#L177) that a server can respond OK to a unary RPC without any detection on the server-side. It'd be nice to throw an exception if close(OK) is called without first sending a message for a UNARY response, although that may be against our API stability. At the very least, we should convert the OK into a Status saying the server is broken.

As reported in grpc/grpc#8010.

### LINKED ISSUE #3059: ServerStreamListenerImpl closing ServerStream in unsafe way
This is not an actual problem because of the way stubs are implemented today, but the code can potentially allow the below to happen:
ServerStreamListenerImpl is executing in the network thread (in the case of the direct executor) or in a application thread (via JumpToApplicationThreadServerStreamListener in the case of an application thread pool). ServerStreamListenerImpl.messageRead can detect an inconsistent gRPC protocol state, and close the ServerStream. This is a problem because the user application responsible for sending data (a different application thread from the JumpToApplicationThreadServerStreamListener) can be in the middle of working with the stream.

A possible fix is to treat this particular invalid gRPC protocol state as a special case, because for clientSendsOne methods, we can require the client to halfClose before we trigger the application code. This allows us to detect the problem in the stub before the application gets triggered, and we avoid any synchronization concerns. We should investigate further into other inconsistent states we want to detect, and make sure this strategy of moving it to the stub also works.

### PR DESCRIPTION
**Pull request by zpencer**
The current check in ServerCallImpl is theoretically unsafe (#3059). Move that check into the stub, and expand the unit tests to cover other interesting edge cases on the server side:
- client sends one, but zero requests received at onHalfClose
- client sends one, but > 1 requests received at onHalfClose
- server sends one, but zero responses sent at onComplete
- server sends one, but > 1 responses sent via onNext

fixes #2243 
fixes #3059

### CODE REVIEW

**Comment by zpencer** [comment]
Synced offline, will move as much validation as possible into ServerCallImpl as possible, and only put things in stubs as a last resort. The reasoning is that the ServerCallImpl is lower level.

**Comment by zpencer** [comment]
retest this please

File: stub/src/main/java/io/grpc/stub/ServerCalls.java [hunk 1]
```diff
@@ -138,8 +157,11 @@ public void onHalfClose() {
                 onReady();
               }
             } else {
-              call.close(Status.INTERNAL.withDescription("Half-closed without a request"),
-                  new Metadata());
+              Status internalError = Status.INTERNAL.withDescription(MISSING_REQUEST);
+              StatusException exception = internalError.asException();
+              log.log(Level.FINE, internalError.getDescription(), exception);
```
Comment (ejona86) [inline stub/src/main/java/io/grpc/stub/ServerCalls.java]: It seems fine to me not to log this, since the client will be told pretty clearly what went wrong and the backtrace in the exception probably isn't too useful.

However, if you are going to log it, don't create the exception unless logging is on. So surround the two lines here with  a `logger.isLoggable(FINE)` check. Ditto elsewhere.

File: core/src/main/java/io/grpc/internal/ServerCallImpl.java [hunk 2]
```diff
@@ -55,6 +63,8 @@
   private boolean sendHeadersCalled;
   private boolean closeCalled;
   private Compressor compressor;
+  private boolean messageSent;
+  private boolean internalClosed;
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerCallImpl.java]: Did you discover `internalClosed` was necessary? I think it should be unnecessary.

File: core/src/main/java/io/grpc/internal/ServerCallImpl.java [hunk 3]
```diff
@@ -156,14 +188,25 @@ public void setCompression(String compressorName) {
...
     closeCalled = true;
-    stream.close(status, trailers);
+    if (internalClosed) {
+      return;
+    }
+
+    if (status.isOk() && method.getType().serverSendsOneMessage() && !messageSent) {
+      internalClose(Status.INTERNAL.withDescription(MISSING_RESPONSE));
+    } else {
+      stream.close(status, trailers);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerCallImpl.java]: Prefer making error cases return to avoid indentation in the common case. Sometimes it doesn't matter, but here it seems like it would improve the readability.

File: core/src/main/java/io/grpc/internal/ServerCallImpl.java [hunk 4]
```diff
@@ -191,6 +234,17 @@ public String getAuthority() {
   }
 
   /**
+   * Close the {@link ServerStream} because an internal error occurred. Allow the application to
+   * run until completion, but silently ignore interactions with the {@link ServerStream} from now
+   * on.
+   */
+  private void internalClose(Status internalError) {
+    internalClosed = true;
+    log.log(Level.FINE, internalError.getDescription(), internalError.asException());
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerCallImpl.java]: I don't think we need to really be logging this. But even if we are, it should be protected with an if isloggable check to avoid creating the exception.

File: core/src/main/java/io/grpc/internal/ServerCallImpl.java (hunk 2 above)
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/ServerCallImpl.java]: The `ServerStream` API states that close() will close the stream for reading and writing, but did not specify what happens if we keep interacting with it after calling close(). `InProcessTransport` and `AbstractStream`  happen to both ignore all interactions after `close`. I didn't want to rely on what seems like behavior, and opted to never touch the stream again after calling `close()`. If you're ok with adding that behavior to the API, this flag can be removed.

File: stub/src/main/java/io/grpc/stub/ServerCalls.java (hunk 1 above)
Comment (zpencer) [inline stub/src/main/java/io/grpc/stub/ServerCalls.java]: removed

File: core/src/main/java/io/grpc/internal/ServerCallImpl.java (hunk 4 above)
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/ServerCallImpl.java]: removed

File: core/src/main/java/io/grpc/internal/ServerCallImpl.java (hunk 3 above)
Comment (zpencer) [inline core/src/main/java/io/grpc/internal/ServerCallImpl.java]: restructured

File: core/src/main/java/io/grpc/internal/ServerCallImpl.java (hunk 2 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerCallImpl.java]: I'm pretty sure we're [relying on similar behavior client-side](https://github.com/grpc/grpc-java/blob/v1.4.0/core/src/main/java/io/grpc/internal/ClientCallImpl.java#L395). I would believe it may be slightly broken, but I'd be in favor of trying to make it a guarantee of the stream.

File: core/src/test/java/io/grpc/internal/ServerCallImplTest.java [hunk 5]
```diff
@@ -159,6 +168,114 @@ public void sendMessage_closesOnFailure() {
...
+    verify(stream, times(1)).writeMessage(any(InputStream.class));
+    verify(stream, times(1)).close(any(Status.class), any(Metadata.class));
+
+    // App runs to completion but everything is ignored
+    serverCall.sendMessage(1L);
+    serverCall.close(Status.OK, new Metadata());
+    try {
+      serverCall.close(Status.OK, new Metadata());
+      fail("calling a second time should still cause an error");
+    } catch (IllegalStateException expected) {
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/ServerCallImplTest.java]: Instead of the try-catch, use `thrown.expect(IllegalStateException.class)` just before calling close the second time?

File: stub/src/main/java/io/grpc/stub/ServerCalls.java [hunk 6]
```diff
@@ -122,24 +129,35 @@ private ServerCalls() {
           ReqT request;
           @Override
           public void onMessage(ReqT request) {
+            if (this.request != null) {
+              // Safe to close the call, because the application has not yet been invoked
```
Comment (ejona86) [inline stub/src/main/java/io/grpc/stub/ServerCalls.java]: invoke() can still happen after this point. It seems you may need a boolean to ignore any future halfClose.

File: stub/src/test/java/io/grpc/stub/ServerCallsTest.java [hunk 7]
```diff
@@ -374,15 +438,19 @@ public Integer parse(InputStream stream) {
   }
 
   private static class ServerCallRecorder extends ServerCall<Integer, Integer> {
-    private List<Integer> requestCalls = new ArrayList<Integer>();
+    private final MethodDescriptor<Integer, Integer> methodDescriptor;
```
Comment (ejona86) [inline stub/src/test/java/io/grpc/stub/ServerCallsTest.java]: This looks unused.

File: stub/src/main/java/io/grpc/stub/ServerCalls.java (hunk 6 above)
Comment (zpencer) [inline stub/src/main/java/io/grpc/stub/ServerCalls.java]: Added the onHalfClose to the test and verified that this now passes the test with the boolean. Also realized that the server streaming tests should really be using the `ServerCalls.asyncServerStreamingCall` stub, even though it's equivalent to `asyncUnaryRequestCall`. I broke the tests into independent tests to be more precise here.

File: core/src/test/java/io/grpc/internal/ServerCallImplTest.java (hunk 5 above)
Comment (zpencer) [inline core/src/test/java/io/grpc/internal/ServerCallImplTest.java]: done

File: stub/src/test/java/io/grpc/stub/ServerCallsTest.java (hunk 7 above)
Comment (zpencer) [inline stub/src/test/java/io/grpc/stub/ServerCallsTest.java]: This field is used in ServerCalls.java:
```
Preconditions.checkArgument(
            call.getMethodDescriptor().getType().clientSendsOneMessage(),
```

This field was previously a "private" and was always null, I turned it into a "private final" and moved it to the rest of the "private final"s.

Comment (ejona86) [inline stub/src/test/java/io/grpc/stub/ServerCallsTest.java]: Ah, it previously being always-null is what caused me to miss the usage.
